### PR TITLE
feature/w-12393084 renewal of invalid sessions

### PR DESCRIPTION
**Pull request by jaizpuru-sf**
# Description

Have the retryTemplate renew session upon INVALID_SESSION_ID errors found 

# Resolves

[W-12393084](https://gus.lightning.force.com/a07EE00001JYIvqYAH)

# Testing

New unit tests added. Pending QA of a SNAPSHOT build on a dev org

## Tests in the PR

- [ ] Manual Tests
- [x] Unit Tests
- [ ] Integration Tests

### CODE REVIEW

**Review by barrerajuanjose — APPROVED** [review]
LGTM!

File: src/main/java/com/sforce/ws/transport/ApiRetryTemplate.java [hunk 1]
```diff
@@ -63,7 +78,18 @@ public <T> T fetchRetrying(FetchOperation<T> fetchOperation, String errorMessage
...
-                if (retriableExClasses.stream().anyMatch(c -> c.isAssignableFrom(e.getClass()))) {
+                if (
+                        this.config != null /* Built to handle session renewals */ &&
+                        e instanceof AsyncApiException &&
+                        ((AsyncApiException) e).getExceptionCode() == AsyncExceptionCode.InvalidSessionId
+                ) {
+                    try {
+                        config.getSessionRenewer().renewSession(config);
+                    } catch (ConnectionException ex) {
+                        throw new AsyncApiException("Cant renew session", AsyncExceptionCode.ClientInputError, ex);
```
Comment (ztaylor797) [inline src/main/java/com/sforce/ws/transport/ApiRetryTemplate.java:89]: "Cant" -> "Can't" and perhaps note that this is referring to a Salesforce session? Just for clarity?

File: src/main/java/com/sforce/ws/transport/ApiRetryTemplate.java [hunk 2]
```diff
@@ -63,7 +78,18 @@ public <T> T fetchRetrying(FetchOperation<T> fetchOperation, String errorMessage
...
                 return fetchOperation.fetch();
             } catch (Exception e) {
-                if (retriableExClasses.stream().anyMatch(c -> c.isAssignableFrom(e.getClass()))) {
+                if (
+                        this.config != null /* Built to handle session renewals */ &&
+                        e instanceof AsyncApiException &&
+                        ((AsyncApiException) e).getExceptionCode() == AsyncExceptionCode.InvalidSessionId
+                ) {
+                    try {
+                        config.getSessionRenewer().renewSession(config);
```
Comment (ztaylor797) [inline src/main/java/com/sforce/ws/transport/ApiRetryTemplate.java:87]: The session renewer is not required for any arbitrary consumer of this lib, so I think we should add a null check here or elsewhere for that case.

**Review by ztaylor797 — CHANGES_REQUESTED** [review]
Left two small comments! Thank you. Nice work.

**Review by ztaylor797 — APPROVED** [review]
LGTM!

**Pull request merged**